crypto/evp/evp_rand.c

```
@@ -0,0 +1,532 @@
+#endif
+       ) {
+        /*
+         * In order to be a consistent set of functions we must have at least
+         * a complete set of "rand" functions and a complete set of context
+         * management functions.  In FIPS mode, we also require the zeroization
+         * verification function.
+         */
+        evp_rand_free(rand);
+        ERR_raise(ERR_LIB_EVP, EVP_R_INVALID_PROVIDER_FUNCTIONS);
+        return NULL;
+    }
+    rand->prov = prov;
+    if (prov != NULL)
+        ossl_provider_up_ref(prov);
+
+    return rand;
+}
+
+EVP_RAND *EVP_RAND_fetch(OPENSSL_CTX *libctx, const char *algorithm,
+                       const char *properties)
+{
+    return evp_generic_fetch(libctx, OSSL_OP_RAND, algorithm, properties,
+                             evp_rand_from_dispatch, evp_rand_up_ref,
+                             evp_rand_free);
+}
+
+int EVP_RAND_up_ref(EVP_RAND *rand)
+{
+    return evp_rand_up_ref(rand);
+}
+
+void EVP_RAND_free(EVP_RAND *rand)
+{
+    evp_rand_free(rand);
+}
+
+int EVP_RAND_number(const EVP_RAND *rand)
+{
+    return rand->name_id;
+}
+
+const char *EVP_RAND_name(const EVP_RAND *rand)
+{
+    return evp_first_name(rand->prov, rand->name_id);
+}
+
+int EVP_RAND_is_a(const EVP_RAND *rand, const char *name)
+{
+    return evp_is_a(rand->prov, rand->name_id, NULL, name);
+}
+
+const OSSL_PROVIDER *EVP_RAND_provider(const EVP_RAND *rand)
+{
+    return rand->prov;
+}
+
+int EVP_RAND_get_params(EVP_RAND *rand, OSSL_PARAM params[])
+{
+    if (rand->get_params != NULL)
+        return rand->get_params(params);
+    return 1;
+}
+
+EVP_RAND_CTX *EVP_RAND_CTX_new(EVP_RAND *rand, EVP_RAND_CTX *parent)
+{
+    EVP_RAND_CTX *ctx;
+    void *parent_ctx = NULL;
+    const OSSL_DISPATCH *parent_dispatch = NULL;
+
+    if (rand == NULL)
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(EVP_RAND_CTX));
+    if (ctx == NULL)
+        return NULL;
+    if (parent != NULL) {
+        EVP_RAND_CTX_enable_locking(parent);
```

should check if locking is possible.  if not, what's the right semantics?  Not needed to lock, or handling locking on your own (which means this code should return an error)

---

Added this check.